pyro/ops/gaussian.py

```diff
@@ -54,7 +110,53 @@ def log_density(self, value):
         result = (-0.5) * torch.matmul(self.precision, value.unsqueeze(-1)).squeeze(-1)
         result = result + self.info_vec
         result = (value * result).sum(-1)
-        return result + self.log_normalizer
+
+        # FIXME is this correct?
+        sqrt_P = self.precision.cholesky(upper=True)
+        sqrt_P_u = self.info_vec.unsqueeze(-1).triangular_solve(sqrt_P).solution.squeeze(-1)
+        u_P_u = sqrt_P_u.pow(2).sum(-1)
+        return result + self.log_normalizer - 0.5 * u_P_u
+
+    def condition(self, value):
+        """
+        Condition this Gaussian on a trailing subset of its state.
+        This should satisfy::
+
+            x.condition(y).dim() == x.dim() - y.size(-1)
+        """
+        assert isinstance(value, torch.Tensor)
+        assert value.size(-1) <= self.info_vec.size(-1)
+
+        # FIXME this is bogus but has the right type:
+        n = self.info_vec.size(-1) - value.size(-1)
+        info_vec = self.info_vec[..., :n]
+        precision = self.precision[..., :n, :n]
+        return Gaussian(self.log_normalizer, info_vec, precision)
+
+    def logsumexp(self):
+        """
+        Integrates out all latent state.
+        """
+        n = self.info_vec.size(-1)
+        return (self.log_normalizer + 0.5 * n * math.log(2 * math.pi) -
+                self.precision.cholesky().diagonal(dim1=-2, dim2=-1).log().sum(-1))
```

Comment:
If we integrate out, we'll need to add `0.5 * info_vec @ inv(precision) @ info_vec` term. Basically, we need to return [the diff term here](https://github.com/pyro-ppl/pyro/blob/dev/pyro/ops/gaussian.py#L105) (you can consider `na = nc = 0` in that implementation.